python/cudaq/dynamics/cudm_solver.py

```diff
@@ -51,21 +53,179 @@ def evolve_dynamics(
     if integrator is None:
         # Default integrator if not provided.
         integrator = RungeKuttaIntegrator()
+
     has_collapse_operators = False
-    if isinstance(hamiltonian, SuperOperator):
-        if len(collapse_operators) > 0:
+
+    if isinstance(hamiltonian, Sequence):
+        # Make sure all Hamiltonians are of the same type.
+        if not all(isinstance(op, Operator) for op in hamiltonian) and not all(
+                isinstance(op, SuperOperator) for op in hamiltonian):
             raise ValueError(
-                "'collapse_operators' must be empty when supplying the super-operator"
+                "All Hamiltonians must be of the same type (either Operator or SuperOperator)"
             )
-        integrator.set_system(dimensions, schedule, hamiltonian)
-        for (left_op, right_op) in hamiltonian:
-            if right_op is not None:
-                has_collapse_operators = True
+        isSuperOperator = isinstance(hamiltonian[0], SuperOperator)
+        if isSuperOperator:
+            if len(collapse_operators) > 0:
+                raise ValueError(
+                    "'collapse_operators' must be empty when supplying a sequence of super-operators"
+                )
+
+            can_be_batched = bindings.checkSuperOpBatchingCompatibility(
+                hamiltonian)
+            batch_size = len(hamiltonian)
+
+            if not can_be_batched:
+                if max_batch_size is not None and max_batch_size > 1:
+                    raise ValueError(
+                        f"The input super-operators are not compatible for batching. Unable to run batched simulation with the requested batch size {max_batch_size}."
+                    )
+                if max_batch_size is None:
+                    print(
+                        "Warning: The input super-operators are not compatible for batching. Running the simulation in non-batched mode."
+                    )
+                # Only run sequentially
+                max_batch_size = 1
+
+            if max_batch_size is None:
```

Comment:
nit: batch_size and this check can be pulled out of the if statement. You could also make a little helper function to check arguments/batch size that takes can_be_batched as argument to avoid the duplication. You could even overload the binding for superoperator and operator and have just a bindings.checkBatchingCompatibility rather than explicitly calling a different function depending on type. Generally, a lot of the code in both branches of the if could be factored out into a helper function to avoid code duplication (also the recursive call).

Reply:
Noted. I'll tidy up this in the upcoming doc PR that I'm working on.

Reply:
Fixed in https://github.com/NVIDIA/cuda-quantum/pull/3129/commits/ba06209eb5c85aee3bc41de834006fad4976bd27